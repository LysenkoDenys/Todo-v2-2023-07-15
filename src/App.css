.App {
  text-align: center;
  max-width: 600px;
  margin: 0 auto;
}

h1 {
  font-size: 35px;
  margin: 10px;
  font-weight: 600;
}

h2 {
  font-size: 16px;
  font-weight: 600;
  margin: 0px;
}
