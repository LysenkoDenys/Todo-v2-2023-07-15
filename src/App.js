import React, { useState } from 'react';
import { v4 as uuidv4 } from 'uuid';
import './App.css';
import TodoForm from './components/Todos/TodoForm';
import TodoList from './components/Todos/TodoList.js';
import TodosActions from './components/Todos/TodosActions';
import Modal from './components/UI/Modal.jsx';

function App() {
  //!====================================
  //get TODO items from localStorage:
  const getStorageItems = () =>
    JSON.parse(localStorage.getItem('textTODO')) || [];
  //!====================================
  const [todos, setTodos] = useState(getStorageItems());
  const [isDescending, setIsDescending] = useState(false);

  const addTodoHandler = (text) => {
    const newTodo = {
      text: `${text} /${new Date(Date.now()).toLocaleString()}`,
      isCompleted: false,
      id: uuidv4(),
    };
    if (text.length) {
      setTodos([...todos, newTodo]);
      //!====================================
      localStorage.setItem('textTODO', JSON.stringify([...todos, newTodo]));
      //!====================================
    }
  };

  const deleteTodoHandler = (id) => {
    setTodos(todos.filter((todo) => id !== todo.id));
    //!====================================
    localStorage.clear();
    localStorage.setItem(
      'textTODO',
      JSON.stringify(todos.filter((todo) => id !== todo.id))
    );
    //!====================================
  };

  const toggleTodoHandler = (id) => {
    setTodos(
      todos.map((todo) =>
        id === todo.id
          ? { ...todo, isCompleted: !todo.isCompleted }
          : { ...todo }
      )
    );
    //!====================================
    localStorage.clear();
    localStorage.setItem(
      'textTODO',
      JSON.stringify(
        todos.map((todo) =>
          id === todo.id
            ? { ...todo, isCompleted: !todo.isCompleted }
            : { ...todo }
        )
      )
    );
    //!====================================
  };

  const resetTodosHandler = () => {
    setTodos([]);
    //!====================================
    localStorage.clear();
    //!====================================
  };

  const deleteCompletedTodosHandler = () => {
    setTodos(todos.filter((todo) => !todo.isCompleted));
    //!====================================
    localStorage.clear();
    localStorage.setItem(
      'textTODO',
      JSON.stringify(todos.filter((todo) => !todo.isCompleted))
    );
    //!====================================
  };

  //!====================================
  const rearrangeTodosHandler = () => {
    setTodos([...todos].reverse());
    setIsDescending((prevIsDescending) => !prevIsDescending);
    localStorage.clear();
    localStorage.setItem('textTODO', JSON.stringify(todos));
  };
  //!====================================

  //!====================================
  const sortDoneTodosHandler = () => {
    setTodos([...todos].sort((a, b) => a.isCompleted - b.isCompleted));
    localStorage.clear();
    localStorage.setItem('textTODO', JSON.stringify(todos));
  };
  //!====================================

  const completedTodosCount = todos.filter((todo) => todo.isCompleted).length;

  return (
    <div className="App">
      <h1>Todo Manager</h1>
      <TodoForm addTodo={addTodoHandler} />
      {!!todos.length && (
        <TodosActions
          completedTodosExist={!!completedTodosCount}
          isDescending={isDescending}
          rearrangeTodos={rearrangeTodosHandler}
          sortDoneTodos={sortDoneTodosHandler}
          resetTodos={resetTodosHandler}
          deleteCompletedTodos={deleteCompletedTodosHandler}
        />
      )}
      {completedTodosCount > 0 && (
        <h2>
          {`You have completed ${completedTodosCount}
            ${completedTodosCount > 1 ? 'todos' : 'todo'}`}
        </h2>
      )}
      <TodoList
        deleteTodo={deleteTodoHandler}
        toggleTodo={toggleTodoHandler}
        todos={todos}
      />
      {/* <Modal /> */}
    </div>
  );
}

export default App;
