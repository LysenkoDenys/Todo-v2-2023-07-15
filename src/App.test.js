import { screen, render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import App from './App.js';

describe('render elements', () => {
  test('render heading, button, input', () => {
    render(<App />);

    const headingElement = screen.getByText(/Todo Manager/i);
    expect(headingElement).toBeInTheDocument();

    const button = screen.getByText(/Submit/i);
    expect(button).toBeInTheDocument();

    const input = screen.getByRole('textbox');
    expect(input).toBeInTheDocument();
    const inputMode = screen.getByPlaceholderText(/Enter new todo/i);
    expect(inputMode).toBeInTheDocument();
    console.log('Actual color:', window.getComputedStyle(button).color);

    expect(input).toMatchSnapshot();
  });
});

test('style', async () => {
  render(<App />);
  const button = await screen.findByText(/Submit/i);
  expect(button).toHaveStyle({
    fontSize: '1.8rem',
    backgroundColor: 'beige',
    fontWeight: '700',
  });
});

// describe('functions', () => {
//   test('functions', () => {
//     render(<App />);
//     expect(resetTodosHandler()).toBe(undefined);
//   });
// });

// describe('events', () => {
//   test('input event', () => {
//     //1. render the component:
//     render(<App />);
//     //2. manipulate the component or find element in it:
//     const input = screen.getAllByRole('textbox');
//     const button = screen.getByText('button');
//     fireEvent.click(button);
//     //3. assertion - make sure the component is doing what we expect it to do:
//     expect(input).toHaveLength(2);
//     expect(button).toBeInTheDocument();
//   });
// });

// npm run test App.test.js
