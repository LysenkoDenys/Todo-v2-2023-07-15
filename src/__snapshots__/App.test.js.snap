// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render elements render heading, button, input 1`] = `
<input
  class="input"
  placeholder="Enter new todo..."
  type="text"
  value=""
/>
`;
